File: quickstart_guides/bit_manipulation/java/PowerOfFour.java
/*
*    Title: Power of four.
*
*    Problem:
*        Given an integer, write a function to determine if it is a power of
*        four.
*
*    Execution: javac PowerOfFour.java && java PowerOfFour
*/
import java.util.*;


public class PowerOfFour {
    public static boolean powerOfFour(int num) {
        if(num==1) return true;
        if(num==0) return false;
        while(num>4 && num%4==0){
            num=num/4;
            }
        if(num==4) return true;
        return false;
    }

    public static void main(String[] args) {
        assert powerOfFour(16) == true;
        assert powerOfFour(5) == false;

        System.out.println("Passed all test cases");
    }
}

File: quickstart_guides/bit_manipulation/python/power_of_four.py
"""
Title: Power of four

Problem: 
    Given an integer, write a function to determine if it is a power of four.

Execution: python power_of_four.py
"""
from math import log
import unittest


def power_of_four(n: int) -> bool:
    return n > 0 and log(n, 4).is_integer()


class TestPowerOfFourr(unittest.TestCase):
    """Unit test for power_of_four."""

    def test_1(self):
        self.assertEqual(power_of_four(16), True)

    def test_2(self):
        self.assertEqual(power_of_four(5), False)

    def test_3(self):
        self.assertEqual(power_of_four(64), True)

    def test_4(self):
        self.assertEqual(power_of_four(18), False)


if __name__ == "__main__":
    unittest.main()
